cpp/src/hdbscan/detail/reachability.cuh

```diff
@@ -162,6 +163,77 @@ void _compute_core_dists(const raft::handle_t& handle,
   core_distances<value_idx>(dists.data(), min_samples, min_samples, m, core_dists, stream);
 }
 
+//  Functor to post-process distances into reachability space
+template <typename T>
+struct ReachabilityPostProcess {
+  void operator()(T* input, size_t batch_i, size_t batch_j, size_t rows, size_t cols) const
+  {
+    // Trying to access member variables directly from the device lambda causes
+    // an invalid memory access for me, copy to a temporary to work
+    // around
+    const T* core_dists_ = core_dists;
+    const T alpha_       = alpha;
+
+    raft::linalg::map_offset(
+      handle, raft::make_device_vector_view(input, rows * cols), [=] __device__(size_t i) {
+        size_t row = i / cols, col = i % cols;
+        return max(core_dists_[col + batch_j], max(core_dists_[row + batch_i], alpha_ * input[i]));
+      });
+  }
+
+  const raft::handle_t& handle;
+  const T* core_dists;
+  T alpha;
+};
+
+/**
+ * Given core distances, Fuses computations of L2 distances between all
+ * points, projection into mutual reachability space, and k-selection.
+ * @tparam value_idx
+ * @tparam value_t
+ * @param[in] handle raft handle for resource reuse
+ * @param[out] out_inds  output indices array (size m * k)
+ * @param[out] out_dists output distances array (size m * k)
+ * @param[in] X input data points (size m * n)
+ * @param[in] m number of rows in X
+ * @param[in] n number of columns in X
+ * @param[in] k neighborhood size (includes self-loop)
+ * @param[in] core_dists array of core distances (size m)
+ */
+template <typename value_idx, typename value_t>
+void mutual_reachability_knn_l2(const raft::handle_t& handle,
+                                value_idx* out_inds,
+                                value_t* out_dists,
+                                const value_t* X,
+                                size_t m,
+                                size_t n,
+                                int k,
+                                value_t* core_dists,
+                                value_t alpha)
+{
+  // Create a functor to postprocess distances into mutual reachability space
+  // Note that we can't use a lambda for this here, since we get errors like:
+  // `A type local to a function cannot be used in the template argument of the
+  // enclosing parent function (and any parent classes) of an extended __device__
+  // or __host__ __device__ lambda`
+  auto post_process = ReachabilityPostProcess<value_t>{handle, core_dists, alpha};
+
+  raft::neighbors::detail::tiled_brute_force_knn(handle,
```

Comment:
We can even set an identity_op for the default epilogue setting here just so we instantiate that by default.